services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/system_monitor/_notifier.py

```diff
@@ -0,0 +1,66 @@
+import contextlib
+import logging
+from pathlib import Path
+
+import socketio
+from fastapi import FastAPI
+from fastapi.encoders import jsonable_encoder
+from models_library.api_schemas_dynamic_sidecar.socketio import (
+    SOCKET_IO_SERVICE_DISK_USAGE_EVENT,
+)
+from models_library.api_schemas_dynamic_sidecar.telemetry import (
+    DiskUsage,
+    ServiceDiskUsage,
+)
+from models_library.projects_nodes_io import NodeID
+from models_library.users import GroupID
+from servicelib.fastapi.app_state import SingletonInAppStateMixin
+
+_logger = logging.getLogger(__name__)
+
+
+class Notifier(SingletonInAppStateMixin):
+    app_state_name: str = "notifier"
+
+    def __init__(self, sio_manager: socketio.AsyncAioPikaManager):
+        self._sio_manager = sio_manager
+
+    async def notify_service_disk_usage(
+        self, primary_group_id: GroupID, node_id: NodeID, usage: dict[Path, DiskUsage]
+    ) -> None:
+        await self._sio_manager.emit(
+            SOCKET_IO_SERVICE_DISK_USAGE_EVENT,
+            data=jsonable_encoder(ServiceDiskUsage(node_id=node_id, usage=usage)),
+            room=f"{primary_group_id}",
```

Comment:
Question: I was wondering why you needed the primary group Id. If this is just to send a message to the user. I would suggest that instead of passing this primary group Id all over the place, just put the user in a user id room as well. This would be so much simpler.  Cause if you do not access the db anywhere there is really no need. And as a matter of fact the dynamic sidecar already accesses the database so I really don't understand the need.

Reply:
I like this proposal of having multiple rooms. We then need to refactor the socket.io code a bit to work both with user_id and gorup_id, since we'd like to avoid sending messages to groups instead of users.


And the reason why the sidecar does not access the db directly, is by design. The only part that access the DB is nodeports. I'd prefer it stayed limited to that. I'd like to avoid using up even more connections.

Reply:
Refactored with this proposal